database/src/backend/heed/database.rs

```diff
@@ -49,6 +49,21 @@ pub(super) struct HeedTableRw<'env, 'tx, T: Table> {
     pub(super) tx_rw: &'tx RefCell<heed::RwTxn<'env>>,
 }
 
+#[allow(clippy::non_send_fields_in_send_ty)] // false positive? <https://github.com/rust-lang/rust-clippy/issues/8045>
+/// SAFETY: 2 invariants for safety:
+///
+/// 1. `cuprate_database`'s Cargo.toml enables a feature
+/// for `heed` that turns on the `MDB_NOTLS` flag for LMDB.
+/// This makes read transactions `Send`, but only if that flag is enabled.
+///
+/// 2. Our `tx_ro` is wrapped in Mutex, as `&T: Send` only if `T: Sync`.
+/// This is what is happening as we have `&TxRw`, not `TxRw`.
+/// <https://github.com/Cuprate/cuprate/pull/113#discussion_r1582189108>
+///
+/// This is required as in `crate::service` we must put our transactions and
+/// tables inside `ThreadLocal`'s to use across multiple threads.
+unsafe impl<T: Table> Send for HeedTableRo<'_, T> {}
```

Comment:
Clippy gives scary warnings without this allow, pretty sure it's a false positive?

Reply:
Only happens after wrapping it in `Mutex` btw, weird.

Reply:
It's because the struct is:
```rust
pub(super) struct HeedTableRo<'tx, T: Table> {
    /// An already opened database table.
    pub(super) db: HeedDb<T::Key, T::Value>,
    /// The associated read-only transaction that opened this table.
    pub(super) tx_ro: Mutex<&'tx heed::RoTxn<'tx>>,
}
```
It needs to be:

 ```rust
 pub(super) struct HeedTableRo<'tx, T: Table> {
    /// An already opened database table.
    pub(super) db: HeedDb<T::Key, T::Value>,
    /// The associated read-only transaction that opened this table.
    pub(super) tx_ro: &'tx Mutex<heed::RoTxn<'tx>>,
}
 ```
 
 Otherwise the mutex isn't actually doing anything. This means you will have to do the same changes that you did to wrap write txs in a refcell :(

Reply:
For future reference:

Discussion in matrix lead to the solution of creating and using a private, separate [`struct UnsafeSendable<T>(T)`](https://github.com/Cuprate/cuprate/pull/113/commits/06a2a869a7d0cc61e115e938578b437173886028) that implements `Send`.

We use it for `HeedTableRo` in [`service`](https://github.com/Cuprate/cuprate/pull/113/commits/08d5715ba96ce663e5c825dee9dcabfe54830d10) where our use-case (inside `ThreadLocal<T: Send>`) is safe.